lib/elixir/lib/calendar/date.ex

```diff
@@ -254,6 +256,12 @@ defmodule Date do
     end
   end
 
+  def from_iso8601(<<?-, next, rest::binary>>, calendar) when next != ?- do
+    with {:ok, %{year: year} = date} <- from_iso8601(<<next>> <> rest, calendar) do
```

Comment:
I think we should extract the parsing to a separate function - right now we'll always do an allocation with the "next" bit.

```elixir
def from_iso8601(<<?-, rest::binary>>, calendar) do
  case parse_iso8601(rest) do
    {year, month, day} -> 
      with {:ok, date} <- new(- year, month, day, Calendar.ISO), do: convert(date, calendar)

    _ -> {:error, :invalid_format}
  end
end

def from_iso8601(string, calendar) do
  case parse_iso8601(rest) do
    {year, month, day} -> 
      with {:ok, date} <- new(year, month, day, Calendar.ISO), do: convert(date, calendar)

    _ -> {:error, :invalid_format}
  end
end

@compile {:inline, parse_iso8601: 1}
defp parse_iso8601(<<year::4-bytes, ?-, month::2-bytes, ?-, day::2-bytes>>) do
  with {year, ""} when year > 0 <- Integer.parse(year),
       {month, ""} <- Integer.parse(month),
       {day, ""} <- Integer.parse(day) do
    {year, month, day}
   end
end
```

Reply:
I went with that "next" allocation for simplicity reason.
I believe negative dates are rare enough to prefer less code over minor optimization.

Reply:
That's fair